streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/GraphGraceSearchUtilTest.java

```diff
@@ -0,0 +1,247 @@
+        final ProcessorParameters<String, Long> processorParameters = new ProcessorParameters<>(processorSupplier, "asdf");
+        final StatefulProcessorNode<String, Long> parentNode = new StatefulProcessorNode<>(
+            "asdf",
+            processorParameters,
+            null,
+            null,
+            false
+        );
+
+        final StatefulProcessorNode<String, Long> statefulNode = new StatefulProcessorNode<>(
+            "stateful",
+            new ProcessorParameters<>(
+                () -> new Processor<Object, Object>() {
+                    @Override
+                    public void init(final ProcessorContext context) {}
+
+                    @Override
+                    public void process(final Object key, final Object value) {}
+
+                    @Override
+                    public void close() {}
+                },
+                "dummy"
+            ),
+            null,
+            null,
+            false
+        );
+        parentNode.addChild(statefulNode);
+
+        final ProcessorGraphNode<String, Long> statelessNode = new ProcessorGraphNode<>("stateless", null);
+        statefulNode.addChild(statelessNode);
+
+        final long extracted = GraphGraceSearchUtil.findAndVerifyWindowGrace(statelessNode);
+        assertThat(extracted, is(windows.gracePeriodMs()));
+    }
+
+    @Test
+    public void shouldExtractGraceFromMultiParentNodes() {
+        final StatefulProcessorNode<String, Long> sessionNode;
+        final long grace = 1234L;
+        {
+            final SessionWindows windows = SessionWindows.with(10L).grace(grace);
+            final KStreamSessionWindowAggregate<String, Long, Integer> processorSupplier =
+                new KStreamSessionWindowAggregate<>(
+                    windows, "asdf", null, null, null
+                );
+            final ProcessorParameters<String, Long> processorParameters = new ProcessorParameters<>(processorSupplier, "asdf");
+            sessionNode = new StatefulProcessorNode<>(
+                "asdf",
+                processorParameters,
+                null,
+                null,
+                false
+            );
+        }
+
+        final StatefulProcessorNode<String, Long> timeNode;
+        {
+            final TimeWindows windows = TimeWindows.of(10L).grace(grace);
+            final KStreamWindowAggregate<String, Long, Integer, TimeWindow> processorSupplier =
+                new KStreamWindowAggregate<>(
+                    windows, "asdf", null, null
+                );
+            final ProcessorParameters<String, Long> processorParameters = new ProcessorParameters<>(processorSupplier, "asdf");
+            timeNode = new StatefulProcessorNode<>(
+                "asdf",
+                processorParameters,
+                null,
+                null,
+                false
+            );
+        }
+
+        final ProcessorGraphNode<String, Long> statelessNode = new ProcessorGraphNode<>("stateless", null);
```

Comment:
The stateless node is the starting point for the search. The two parents are stateful nodes because the test needs the parents to be windowed aggregations, which are stateful.

The child node could be stateful as well, but it's not necessary.

Do you think this test should be duplicated to make sure the search works using both a stateless and a stateful processing code as the starting point?

Reply:
I understood the setup -- the question is, is there any difference in stateful/stateless parent that is worth testing? If not, we should use the same for all tests (otherwise, it's confusion --- at least to me -> "why do you need to use the one or the other?" question arrises)